### ISSUE TITLE
about VAE model

### ISSUE THREAD

**Issue by yangXJU** [masato-ka/airc-rl-agent#24]
hi
I am using LEGO city raods, but reconstruction image is not so good, I notice the VAE model is as follow
Does 3 represent the number of input channels and 32 represent the number of output channels?
And do you think I should train VAE myself?

VAE(
  (encoder): Sequential(
    (0): Conv2d(3, 32, kernel_size=(4, 4), stride=(2, 2))
    (1): ReLU()
    (2): Conv2d(32, 64, kernel_size=(4, 4), stride=(2, 2))
    (3): ReLU()
    (4): Conv2d(64, 128, kernel_size=(4, 4), stride=(2, 2))
    (5): ReLU()
    (6): Conv2d(128, 256, kernel_size=(4, 4), stride=(2, 2))
    (7): ReLU()
    (8): Flatten()
  )
  (fc1): Linear(in_features=6144, out_features=32, bias=True)
  (fc2): Linear(in_features=6144, out_features=32, bias=True)
  (fc3): Linear(in_features=32, out_features=6144, bias=True)
  (decoder): Sequential(
    (0): UnFlatten()
    (1): ConvTranspose2d(256, 128, kernel_size=(4, 4), stride=(2, 2))
    (2): ReLU()
    (3): ConvTranspose2d(128, 64, kernel_size=(4, 4), stride=(2, 2))
    (4): ReLU()
    (5): ConvTranspose2d(64, 32, kernel_size=(5, 5), stride=(2, 2))
    (6): ReLU()
    (7): ConvTranspose2d(32, 3, kernel_size=(4, 4), stride=(2, 2))
    (8): Sigmoid()
  )
)

**Comment by masato-ka** [comment]
If you can not get good reconstruction image, recommend to re-training VAE model. You can use below colab notebook for that.

https://github.com/masato-ka/airc-rl-agent/blob/master/notebooks/colabo/VAE_CNN.ipynb

But i think, for state representation learning, not important quality of reconstruction image. Consistent Latent spaces is important. you can check latent spaces by using below notebook. 

https://github.com/masato-ka/airc-rl-agent/blob/master/notebooks/utility/jetbot/vae_viewer.ipynb

>Does 3 represent the number of input channels and 32 represent the number of output channels?

Yes, the first layer Conv2D is that mean. Input(3) is image channel(RGB). output(32) is filter of Conv2D. 
Please refer to Pytorch API documentation.

**Comment by masato-ka** [comment]
Is your question solved? In that case, I will close this issue tomorrow.

**Comment by yangXJU** [comment]
I'm very sorry to reply so late. It's been a busy time recently. I need to train on my own equipment, so I need to modify the VAE training code, which may take a day.

**Comment by yangXJU** [comment]
hi, 
I got a size mismatch m1:[64*59904] , m2:[6144*32]
after debug, the error is caused by

line 173 :  h = self.encoder(x)
line 175 :  z, mu, logvar = self.bottleneck(h)
line  168 : mu, logvar = self.fc1(h), F.softplus(self.fc2(h))

I guess fc1 require a tensor which is 64*6144, but h is 64*59904

the pictures I use for training are 320*240, and I run VAE_CNN on my computer with a 1080Ti
any advice

**Comment by masato-ka** [comment]
This is your original implementation ? I don't know your model parameter. But If you use Pytorch, you can use torchsummary for debug.

For example.
```
from torchsummary import summary
vae = VAE(image_channels=image_channels, z_dim=VARIANTS_SIZE ).to(device)
summary(vae, (3, 80, 160))
```

**Comment by Adnan-annan** [comment]
hi @yangXJU  did you manage to train the VAE model yourself ?

**Comment by masato-ka** [comment]
I will close in 3 days when do not update.
